libclc/amdgcn-amdhsa/libspirv/math/frexp.inc

```diff
@@ -41,24 +48,68 @@ __spirv_ocl_frexp(__CLC_GENTYPE_VEC x, __CLC_ADDRESS_SPACE int3 *ep) {
   return res;
 }
 
-#define __CLC_VECTORIZE_FREXP_IMPL(GENTYPE, VEC_LEN, HALF_VEC_LEN)             \
-  _CLC_OVERLOAD _CLC_DEF GENTYPE __spirv_ocl_frexp(                            \
-      GENTYPE x, __CLC_ADDRESS_SPACE int##VEC_LEN *ep) {                       \
-    int##HALF_VEC_LEN ep_lo;                                                   \
-    int##HALF_VEC_LEN ep_hi;                                                   \
-    GENTYPE res = (GENTYPE)(__spirv_ocl_frexp(x.lo, &ep_lo),                   \
-                            __spirv_ocl_frexp(x.hi, &ep_hi));                  \
-    *ep = (int##VEC_LEN)(ep_lo, ep_hi);                                        \
-    return res;                                                                \
-  }
+#undef __CLC_GENTYPE_VEC
+#define __CLC_GENTYPE_VEC __CLC_XCONCAT(__CLC_GENTYPE, 4)
+
+_CLC_DEF __CLC_GENTYPE_VEC VEC_FUNCNAME(4)(__CLC_GENTYPE_VEC x,
+                                           int4 __CLC_ADDRESS_SPACE *ep) {
+  int ep_0, ep_1, ep_2, ep_3;
+  __CLC_GENTYPE_VEC res = (__CLC_GENTYPE_VEC)(__spirv_ocl_frexp(x.s0, &ep_0),
+                                              __spirv_ocl_frexp(x.s1, &ep_1),
+                                              __spirv_ocl_frexp(x.s2, &ep_2),
+                                              __spirv_ocl_frexp(x.s3, &ep_3));
```

Comment:
Can't we use `VEC_FUNCNAME(2)` here? Similar to the existing approach where computation for vector length 2*N done by combining results for vector length N.

Reply:
Yes, that is possible but would it not lead to extra vector being created? For example VEC_FUNCNAME(16) would create a 8 int2, 4 int4, 2 int8 and an int16.

Reply:
VEC_FUNCNAME(16) should create just 2 int8. Currently it creates 16(!) scalar values, so I don't think creating 2 int8 should be a problem.

Reply:
Sorry, I think I didn't get your concern right. There should be no overhead on creating vectors. Compiler should generate the same for code for both variants. It's just less code to read.

Reply:
Ok, looking at _CLC_V_V_VP_VECTORIZE and modf this also seems to be a fairly common pattern. Thanks!